File: src/PlaceholdersUtils.php
<?php

declare(strict_types=1);

namespace Keboola\GenericExtractor;

use Keboola\Code\Builder;
use Keboola\GenericExtractor\Exception\UserException;
use Keboola\Utils\Exception\NoDataFoundException;
use function Keboola\Utils\arrayToObject;
use function Keboola\Utils\getDataFromPath;

class PlaceholdersUtils
{
    public static function getParamsForChildJobs(array $placeholders, array $parentResults, array $parentParams): array
    {
        $params = [];
        foreach ($placeholders as $placeholder => $field) {
            $params[$placeholder] = self::getPlaceholder($placeholder, $field, $parentResults);
        }

        // Add parent params as well (for 'tagging' child-parent data)
        // Same placeholder in deeper nesting replaces parent value
        $params = array_replace($parentParams, $params);

        // Create all combinations if there are some parameter values as array.
        // Each combination will be one child job.
        return self::flattenParameters($params);
    }

    /**
     * @param  string|array $field Path or a function with a path
     * @return array ['placeholder', 'field', 'value']
     */
    public static function getPlaceholder(string $placeholder, $field, array $parentResults): array
    {
        // TODO allow using a descriptive ID(level) by storing the result by `task(job) id` in $parentResults
        $level = strpos($placeholder, ':') === false
            ? 0
            : (int) strtok($placeholder, ':') -1;

        // Check function (defined as array)
        if (!is_scalar($field)) {
            if (empty($field['path'])) {
                throw new UserException(
                    "The path for placeholder '{$placeholder}' must be a string value or an object ".
                    "containing 'path' and 'function'."
                );
            }

            $fn = (object) arrayToObject($field);
            $field = $field['path'];
            unset($fn->path);
        }

        // Get value
        $value = self::getPlaceholderValue($field, $parentResults, $level, $placeholder);

        // Run function
        if (isset($fn)) {
            $builder = new Builder;
            $builder->allowFunction('urlencode');
            $value = $builder->run($fn, ['placeholder' => ['value' => $value]]);
        }

        // Return definition
        return [
            'placeholder' => $placeholder,
            'field' => $field,
            'value' => $value,
        ];
    }

    /**
     * @return mixed
     */
    public static function getPlaceholderValue(string $field, array $parentResults, int $level, string $placeholder)
    {
        try {
            if (!array_key_exists($level, $parentResults)) {
                $maxLevel = empty($parentResults) ? 0 : (int) max(array_keys($parentResults)) +1;
                throw new UserException(
                    'Level ' . ++$level . ' not found in parent results! Maximum level: ' . $maxLevel
                );
            }

            return getDataFromPath($field, $parentResults[$level], '.', false);
        } catch (NoDataFoundException $e) {
            throw new UserException(
                "No value found for {$placeholder} in parent result. (level: " . ++$level . ')',
                0,
                null,
                [
                    'parents' => $parentResults,
                ]
            );
        }
    }

    public static function flattenParameters(array $params): array
    {
        $flatParameters = [];
        $i = 0;
        foreach ($params as $placeholderName => $placeholder) {
            $template = $placeholder;
            if (is_array($placeholder['value'])) {
                foreach ($placeholder['value'] as $value) {
                    $template['value'] = $value;
                    $flatParameters[$i][$placeholderName] = $template;
                    $i++;
                }
            } else {
                $flatParameters[$i][$placeholderName] = $template;
            }
        }
        return $flatParameters;
    }
}

File: python-sync-actions/src/placeholders_utils.py
from collections import namedtuple
from itertools import product
from typing import List, Dict, Union, Any


class UserException(Exception):
    pass


class NoDataFoundException(Exception):
    pass


Placeholder = namedtuple('Placeholder', ['placeholder', 'json_path', 'value'])


class PlaceholdersUtils:

    @staticmethod
    def get_params_for_child_jobs(placeholders: Dict[str, Any], parent_results: List[Dict[str, Any]],
                                  parent_params: Dict[str, Any]) -> List[Dict[str, Any]]:
        params = {}
        for placeholder, field in placeholders.items():
            params[placeholder] = PlaceholdersUtils.get_placeholder(placeholder, field, parent_results)

        # Add parent params as well (for 'tagging' child-parent data)
        # Same placeholder in deeper nesting replaces parent value
        params = {**parent_params, **params}

        # Create all combinations if there are some parameter values as array.
        # Each combination will be one child job.
        return PlaceholdersUtils.get_params_per_child_job(params)

    @staticmethod
    def get_placeholder(placeholder: str, field: Union[str, Dict[str, Any]],
                        parent_results: List[Dict[str, Any]]) -> Dict[str, Any]:
        # Determine the level based on the presence of ':' in the placeholder name
        level = 0 if ':' not in placeholder else int(placeholder.split(':')[0]) - 1

        # Check function (defined as dict)
        if not isinstance(field, str):
            if 'path' not in field:
                raise UserException(f"The path for placeholder '{placeholder}' must be a string value"
                                    f"or an object containing 'path' and 'function'.")
            fn = field.copy()
            field = fn.pop('path')

        # Get value
        value = PlaceholdersUtils.get_placeholder_value(str(field), parent_results, level, placeholder)

        # Run function if provided
        if 'fn' in locals():
            # Example function to be replaced by actual implementation
            value = value

        return {
            'placeholder': placeholder,
            'json_path': field,
            'value': value
        }

    @staticmethod
    def get_placeholder_value(field: str, parent_results: List[Dict[str, Any]], level: int, placeholder: str) -> Any:
        try:
            if level >= len(parent_results):
                max_level = 0 if not parent_results else len(parent_results)
                raise UserException(f'Level {level + 1} not found in parent results! Maximum level: {max_level}')

            # Implement get_data_from_path to fetch data using a dot notation
            data = get_data_from_path(field, parent_results[level])

            return data

        except NoDataFoundException:
            raise UserException(
                f"No value found for placeholder {placeholder} in parent result. (level: {level + 1})",
                None, None, {'parents': parent_results}
            )

    @staticmethod
    def get_params_per_child_job(params: Dict[str, Placeholder]) -> List[Dict[str, Any]]:
        # Flatten parameters to a list of lists
        flattened = {}
        for placeholder_name, placeholder in params.items():
            if isinstance(placeholder['value'], list):
                flattened[placeholder_name] = [
                    {'placeholder': placeholder_name, 'json_path': placeholder['json_path'], 'value': value}
                    for value in placeholder['value']
                ]
            else:
                flattened[placeholder_name] = [placeholder]

        # Cartesian product to get all combinations
        return PlaceholdersUtils.cartesian(flattened)

    @staticmethod
    def cartesian(input_data: Dict[str, List[Dict[str, Any]]]) -> List[Dict[str, Any]]:
        # Generate the Cartesian product of input lists
        keys, values = zip(*input_data.items())
        product_list = [dict(zip(keys, combination)) for combination in product(*values)]

        return product_list


def get_data_from_path(json_path: str, data: Dict[str, Any], separator: str = '.', strict: bool = True) -> Any:
    """Mock function to fetch data using a dot-separated path notation. Replace with actual implementation."""
    keys = json_path.split(separator)
    for key in keys:
        if key not in data:
            if strict:
                raise NoDataFoundException(f"Key '{key}' not found in login data.")
            return None
        data = data[key]
    return data
